### PR TITLE
Borrowing in dart

### PR DESCRIPTION
**Pull request by Manishearth**
This adds support for retaining ownership edges when borrowing across FFI in Dart, using the HIR lifetimes support. I've tried to have as much of the complicated lifetime logic live inside the HIR module. It still ends up being somewhat complicated on the tool side, but not nearly as complicated as the JS stuff.

A design decision made here is to _not_ recurse into struct fields when figuring out lifetime stuff for a method; we instead call a method on the struct that gives us what we need, called `_fields_for_lifetime_foo()` which appends every field that could potentially borrow from `'foo`.

This unfortunately gets awkward for slices in structs: for everything else the fields are borrowing from the lifetime, but slices are not converted until later when we call `struct._pointer()`, and it all gets stuffed into the same arena that gets freed later. I haven't yet solved this, but I would like to land this code first. I see a couple implementation strategies for this, explained in a comment below. I'd like to land this first.

A thing I am definitely planning to do (but did not include in this PR) is to get rid of the TypeLifetimes vs MethodLifetimes distinction: it _is_ useful to keep track of what the "outer" and "inner" lifetimes are but the problem is that it can't happen at the type level: the lifetimes on a `Type<P>` can be method or type lifetimes depending on whether the `Type<P>` is found inside a struct def or method. Reviewers, ignore any ugliness caused by that distinction.

### CODE REVIEW

**Comment by Manishearth** [comment]
@robertbastian you can start reviewing this if you'd like, this is partial but it's mergeable

The method stuff is complicated and will take a bit more time

**Comment by Manishearth** [comment]
@robertbastian I've got a lot of changes locally building on this (some of which undoes previous work here), I will likely force push over your commit

**Comment by robertbastian** [comment]
ok

**Comment by Manishearth** [comment]
@robertbastian ready for review. Happy to talk through the complexities.

This doesn't handle slices correctly, and it's tricky with the current design. I'd like to land this first, though.

I see three paths forward for slices:




 - `.pointer()` accepts lifetime edge arrays and appends the relevant arenas to them.
    - Upside: Non-slice methods work smoothly, it's only slice stuff that's weird
    - Downside: Slices get handled differently from everything else. Less clean
 - We no longer initialize the method edge arrays, and instead each parameter appends to the relevant ones. They get passed down to `.pointer()`. We avoid doing extraneous stuff by passing null if an edge array isn't needed.
   - Upside: uniform treatment for everyone. Also, nice clean symmetry of structs handling their borrowing in their constructor and `pointer()`
   - Downside: instead of statically constructing an edge array, we're dynamically constructing them. Not great for something that ideally affects performance minimally.
 - we abandon the approach where each type only manages its own borrowing and have methods recursively do everything for struct parameters
   - Upside: You can see exactly which fields are being borrowed.
   - Downside: code duplication, code is harder to read 


In the first two approaches, instead of having one arena to rule them all, _for borrowed fields only_ we construct individual arenas that have a finalizer attached.


Less precise approaches would be to always make a single arena, but if we know it is being borrowed from append it to every relevant lifetime edge and switch it to using a finalizer.

**Comment by Manishearth** [comment]
My hope is to iterate on this as much as possible and then once we're happy with it refactor stuff to mostly live in the hir module in a reusable way

**Review by robertbastian — COMMENTED** [review]
The big problem is that Arenas don't release their memory when dropped, and it's not even guaranteed that they will ever get dropped. I'll have to talk to Daco about how to best clean these up.

**Comment by Manishearth** [comment]
@robertbastian I was thinking that we can make a SliceEdge type with a finalizer that frees the contained arena.

(yes, finalizers aren't guaranteed, but that's part and parcel of using a GCd language. We're not doing anything that would make finalizers not run)

**Review by robertbastian — APPROVED** [review]
Look good so far

File: tool/src/dart/mod.rs [hunk 1]
```diff
@@ -330,6 +334,100 @@ impl<'a, 'cx> TyGenContext<'a, 'cx> {
...
+                    // since the _fields_for_lifetime_foo() methods are named after
+                    // the *def* context lifetime.
+                    //
+                    // Concretely, if we have struct `Foo<'a, 'b>` and our method
+                    // accepts `Foo<'x, 'y>`, we need to output _fields_for_lifetime_a()/b not x/y.
+                    let def_lifetimes = def.lifetimes.all_lifetimes();
+                    let use_lifetimes = s.lifetimes.lifetimes();
+                    assert_eq!(def_lifetimes.len(), use_lifetimes.len(), "lifetimes array found on struct def must match lifetime parameters accepted by struct");
+                    // the type lifetimes array
+                    for (def_lt, use_lt) in def_lifetimes.zip(use_lifetimes) {
```
Comment (Manishearth) [inline tool/src/dart/mod.rs:390]: thought: this trick works for structs but not for borrowing opaques, which can have a `self` lifetime in `use_lifetimes`

**Pull request merged**